crates/typst/src/layout/grid/layout.rs

```diff
@@ -347,23 +82,225 @@ impl<T: FromValue> FromValue for Celled<T> {
+    ) -> Cell;
+
+    /// Creates a cell with empty content.
+    /// Needed to fill incomplete rows.
+    fn new_empty_cell() -> Self;
+}
+
+/// A grid of cells, including the columns, rows, and cell data.
+pub struct CellGrid {
     /// The grid cells.
-    cells: &'a [Content],
-    /// Whether this is an RTL grid.
-    is_rtl: bool,
-    /// Whether this grid has gutters.
-    has_gutter: bool,
+    cells: Vec<Cell>,
     /// The column tracks including gutter tracks.
     cols: Vec<Sizing>,
     /// The row tracks including gutter tracks.
     rows: Vec<Sizing>,
-    // How to fill the cells.
-    #[allow(dead_code)]
-    fill: &'a Celled<Option<Paint>>,
+    /// Whether this grid has gutters.
+    has_gutter: bool,
+    /// Whether this is an RTL grid.
+    is_rtl: bool,
+}
+
+impl CellGrid {
+    /// Generates the cell grid, given the tracks and resolved cells.
+    pub fn new(
+        tracks: Axes<&[Sizing]>,
+        gutter: Axes<&[Sizing]>,
+        cells: Vec<Cell>,
+        styles: StyleChain,
+    ) -> Self {
+        let mut cols = vec![];
+        let mut rows = vec![];
+
+        // Number of content columns: Always at least one.
+        let c = tracks.x.len().max(1);
+
+        // Number of content rows: At least as many as given, but also at least
+        // as many as needed to place each item.
+        let r = {
+            let len = cells.len();
+            let given = tracks.y.len();
+            let needed = len / c + (len % c).clamp(0, 1);
+            given.max(needed)
+        };
+
+        let has_gutter = gutter.any(|tracks| !tracks.is_empty());
+        let auto = Sizing::Auto;
+        let zero = Sizing::Rel(Rel::zero());
+        let get_or = |tracks: &[_], idx, default| {
+            tracks.get(idx).or(tracks.last()).copied().unwrap_or(default)
+        };
+
+        // Collect content and gutter columns.
+        for x in 0..c {
+            cols.push(get_or(tracks.x, x, auto));
+            if has_gutter {
+                cols.push(get_or(gutter.x, x, zero));
+            }
+        }
+
+        // Collect content and gutter rows.
+        for y in 0..r {
+            rows.push(get_or(tracks.y, y, auto));
+            if has_gutter {
+                rows.push(get_or(gutter.y, y, zero));
+            }
+        }
+
+        // Remove superfluous gutter tracks.
+        if has_gutter {
+            cols.pop();
+            rows.pop();
+        }
+
+        // Reverse for RTL.
+        let is_rtl = TextElem::dir_in(styles) == Dir::RTL;
+        if is_rtl {
+            cols.reverse();
+        }
+
+        Self { cols, rows, cells, has_gutter, is_rtl }
+    }
+
+    /// Resolves all cells in the grid before creating it.
+    /// Allows them to keep track of their final properties and adjust their fields accordingly.
+    #[allow(clippy::too_many_arguments)]
+    pub fn resolve<T: ResolvableCell>(
+        tracks: Axes<&[Sizing]>,
+        gutter: Axes<&[Sizing]>,
+        cells: Vec<T>,
```

Comment:
could this be `IntoIterator<Item = T>`?

Reply:
Theoretically, in the current form, yes; however, we lose the ability to predict the cell vector's length, which is eventually going to be necessary for proper optimization (unless we can use the iterator size hint thing?). And we only use vectors in all current usages of CellGrid anyways. But if we can find an alternative way to obtain the amount of cells, then that can work.

Reply:
From a quick look those vectors wouldn't need to be cloned if this was an iterator (just their items, i.e. `.iter().cloned()`). It could also, as said, use the size hint or even ExactSizeIterator. But it's not important, just a suggestion.

Reply:
I've changed it to take a slice now. Should probably be simpler, and for now should give us the same benefits of not taking an owned vector. We can change later if needed.